### ISSUE TITLE
NSwag generates incorrect serialization for date parameters

### ISSUE THREAD

**Issue by williamyaoh-ah** [RSuter/NSwag#1211]
According to the Swagger 2.0 spec, ["type": "string", "format": "date" should](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#data-types) be formatted according to `full-date` [here](https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14). (E.g. "2018-01-01") However, even when a request parameter is specified this way, NSwag serializes the date with a timestamp appended, like "2018-01-03T00:00:00".

Here's a minimal Swagger specification that exhibits this behavior:

```
{
    "swagger": "2.0",
    "info": {
        "version": "0.0.1.0",
        "title": "API"
    },
    "paths": {
        "/api/endpoint": {
            "post": {
                "consumes": [ "application/json" ],
                "parameters": [
                    { "name": "date",
                      "in": "body",
                      "schema": {
                          "type": "string",
                          "format": "date"
                      }
                    }
                ],
                "responses": {
                    "200": {
                    }
                }
            }
        }
    }
}
```

After generating a client from this specification using NSwag -> C# and running it against a local server, the parameter has a timestamp attached to the date.

**Comment by RSuter** [comment]
Good point, we should probably generate a custom json converter and add it to the date properties.

https://stackoverflow.com/a/18639829/876814

Do you know if there is a way to correctly serialize these dates without having to generate the custom json converter?

**Comment by williamyaoh-ah** [comment]
> Do you know if there is a way to correctly serialize these dates without having to generate the custom json converter?

Not that I know of. With the built-in date conversion base class in Newtonsoft, it doesn't seem like that much extra generated code to add though.

I remember having issues dealing with nested DateTimes when trying to customize Newtonsoft's JSON serialization in the past though; let me see if I can repro that, cuz getting *all* usages of a `date` formatted string in Swagger might be more complicated than at first glance.

**Comment by williamyaoh-ah** [comment]
Yup, if the DateTimes are inside another data structure, things get ugly. Here's some example C# where the dates are in a List; compiles fine, errors at runtime:

```csharp
using System;
using System.Collections.Generic;

using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace Program {
  public class CustomDateTimeConverter : IsoDateTimeConverter {
    public CustomDateTimeConverter() {
      base.DateTimeFormat = "yyyy-MM-dd";
    }
  }

  public class JSON1 {
    [JsonConverter( typeof(CustomDateTimeConverter) )]
    [JsonProperty( "date" )]
    public DateTime Date { get; set; }
  }

  public class JSON2 {
    [JsonConverter( typeof(CustomDateTimeConverter) )]
    [JsonProperty( "dates" )]
    public List<DateTime> Dates { get; set; }
  }

  public class Program {
    public static void Main(String[] args) {
      var json1 = new JSON1() { Date = new DateTime(2018, 01, 03) };
      var json2 = new JSON2() {
        Dates = new List<DateTime> {
          new DateTime(2018, 01, 03),
          new DateTime(2018, 01, 17),
          new DateTime(2018, 01, 31)
        }
      };

      Console.WriteLine(JsonConvert.SerializeObject(json1));
      Console.WriteLine(JsonConvert.SerializeObject(json2));
    }
  }
}
```

So the nice, simple converter can't be used, say, when the Swagger spec contains an array of `date`-formatted strings, and another, more complicated converter would be needed for that case. Ugh.

There is [this example](https://www.newtonsoft.com/json/help/html/SerializeDateFormatString.htm) of serializing a date by specifying the `DateFormatString` directly in the serializer settings. Unfortunately, that would have collision problems if a Swagger spec contained both dates without timestamps and dates with timestamps...

**Comment by RSuter** [comment]
Yes, it should be a property "local" converter and not a global converter. Maybe we can implement a whole new converter which is not based IsoDateTimeConverter and can handle date, date array and date dictionary properties?

**Comment by williamyaoh-ah** [comment]
Hmm... but that wouldn't work if there were multiple nested collections; for example, maybe someone wants to represent date *ranges*, and uses an array of two-element subarrays, each containing a date. In the general recursive case, the level of possible nesting is infinite, and we'd need an infinite amount of generated converters... obviously, in any given specification, there's only going to be a finite amount of expressed nestings, but still. Having to recursively generate the code in each converter feels... wrong.

Grr, dammit Newtonsoft. There should really just be a "here's a converter for a given type; within this class member, recursively serialize any instances of this type using this converter" property.

**Comment by RSuter** [comment]
Yes, but we could implement a recursive converter? :-) Or we just support the most common scenarios...

**Comment by zlamma** [comment]
There's the same problem with dateTime.

[OpenAPI `dateTime`](https://swagger.io/specification/#data-types-13) should be formatted according to [RFC 3339 `date-time`](https://xml2rfc.tools.ietf.org/public/rfc/html/rfc3339.html#anchor14) which, if you look closely, **makes the timezone mandatory**.

Sadly, the generated DTOs and controller classes accept values without timezone just fine.

I have tried suggesting making it easier for every ASP.NET based framework by [implementing this on Newtonsoft.JSON](https://github.com/JamesNK/Newtonsoft.Json/issues/1631) but no success for now. The lib author indeed [suggests writing a date time converter](https://github.com/JamesNK/Newtonsoft.Json/issues/1631#issuecomment-372183087) (I've [asked](https://github.com/JamesNK/Newtonsoft.Json/issues/1631#issuecomment-374246618) whether a PR would be considered).

**Comment by RSuter** [comment]
I think we can address both here...

**Comment by andi0b** [comment]
@RicoSuter: I would consider this a major bug, not an enhancement, as it violates the OpenAPI spec and there is no workaround which doesn't break date-time parameters. 

I guess there should also be an additional config paramter `parameterDateFormat`  besides `parameterDateTimeFormat` to allow "customization" which is compliant to OpenAPI:

```c#
 // for "date"
DateTime.Now.ToString("yyyy-MM-dd");

// for "date-time"
DateTime.Now.ToString("o"); 
```

This should also be possible in JSON-Serialization, but that's another issue I guess.

**Comment by knoepdan** [comment]
We have the same problem. (See   #2084  ) 
An additional parameter "parameterDateFormat" would indeed solve the problem.
